
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../installation/">
      
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.27">
    
    
      
        <title>Genotype-Corrector Tutorial - The Schnable Lab@UNL</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../mkapi-common.css">
    
      <link rel="stylesheet" href="../mkapi-material.css">
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../css/termynal.css">
    
      <link rel="stylesheet" href="../css/custom.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#step-1-prepare-genotype-data-to-the-required-data-format" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="The Schnable Lab@UNL" class="md-header__button md-logo" aria-label="The Schnable Lab@UNL" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            The Schnable Lab@UNL
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Genotype-Corrector Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18c-.89 0-1.74-.2-2.5-.55C11.56 16.5 13 14.42 13 12c0-2.42-1.44-4.5-3.5-5.45C10.26 6.2 11.11 6 12 6a6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12 20 8.69Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/chenyong89/unl-schnablelab" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    unl-schnablelab
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="./" class="md-tabs__link">
        
  
    
  
  Genotype-Corrector Tutorial

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="The Schnable Lab@UNL" class="md-nav__button md-logo" aria-label="The Schnable Lab@UNL" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    The Schnable Lab@UNL
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/chenyong89/unl-schnablelab" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    unl-schnablelab
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Genotype-Corrector Tutorial
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Genotype-Corrector Tutorial
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#step-1-prepare-genotype-data-to-the-required-data-format" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1: Prepare Genotype Data to the Required Data Format
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-2-quality-control-qc-on-genotype-data" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2: Quality Control (QC) on Genotype Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 2: Quality Control (QC) on Genotype Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qc-on-missing-data" class="md-nav__link">
    <span class="md-ellipsis">
      QC on Missing Data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc-on-snps-with-distorted-segregation-ratio" class="md-nav__link">
    <span class="md-ellipsis">
      QC on SNPs with Distorted Segregation Ratio
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc-on-homozygous-genotype-calls-in-heterozygous-regions" class="md-nav__link">
    <span class="md-ellipsis">
      QC on Homozygous Genotype Calls in Heterozygous Regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc-on-duplicated-snp-markers" class="md-nav__link">
    <span class="md-ellipsis">
      QC on Duplicated SNP Markers
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-3-conduct-the-correction" class="md-nav__link">
    <span class="md-ellipsis">
      Step 3. Conduct the Correction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 3. Conduct the Correction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prepare-the-configuration-file" class="md-nav__link">
    <span class="md-ellipsis">
      Prepare the Configuration File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-genotype-corrector" class="md-nav__link">
    <span class="md-ellipsis">
      Run Genotype-Corrector
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tips" class="md-nav__link">
    <span class="md-ellipsis">
      Tips
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#step-1-prepare-genotype-data-to-the-required-data-format" class="md-nav__link">
    <span class="md-ellipsis">
      Step 1: Prepare Genotype Data to the Required Data Format
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-2-quality-control-qc-on-genotype-data" class="md-nav__link">
    <span class="md-ellipsis">
      Step 2: Quality Control (QC) on Genotype Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 2: Quality Control (QC) on Genotype Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#qc-on-missing-data" class="md-nav__link">
    <span class="md-ellipsis">
      QC on Missing Data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc-on-snps-with-distorted-segregation-ratio" class="md-nav__link">
    <span class="md-ellipsis">
      QC on SNPs with Distorted Segregation Ratio
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc-on-homozygous-genotype-calls-in-heterozygous-regions" class="md-nav__link">
    <span class="md-ellipsis">
      QC on Homozygous Genotype Calls in Heterozygous Regions
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#qc-on-duplicated-snp-markers" class="md-nav__link">
    <span class="md-ellipsis">
      QC on Duplicated SNP Markers
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#step-3-conduct-the-correction" class="md-nav__link">
    <span class="md-ellipsis">
      Step 3. Conduct the Correction
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Step 3. Conduct the Correction">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#prepare-the-configuration-file" class="md-nav__link">
    <span class="md-ellipsis">
      Prepare the Configuration File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#run-genotype-corrector" class="md-nav__link">
    <span class="md-ellipsis">
      Run Genotype-Corrector
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#tips" class="md-nav__link">
    <span class="md-ellipsis">
      Tips
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Genotype-Corrector Tutorial</h1>

<h2 id="step-1-prepare-genotype-data-to-the-required-data-format">Step 1: Prepare Genotype Data to the Required Data Format</h2>
<p>The input genotype table is a tab-delimited text file containing genotypes for each sample/line in a population. Markers must be sorted by position within each chromosome or contig. A typical input genotype table format for GC should resemble the following:
<img alt="input_data_format" src="../figures/data_format.png" /></p>
<p>If your genotype data is in VCF format, you can use the following command-line tool to convert it to the required format:
<div class="language-text highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>$ python -m schnablelab.GC.data_qc vcf2map input.vcf output.map
</code></pre></div></p>
<h2 id="step-2-quality-control-qc-on-genotype-data">Step 2: Quality Control (QC) on Genotype Data</h2>
<p>Several QC steps are necessary before proceeding to the correction step. Example data for this demo can be found in the repository <a href="https://github.com/chenyong89/unl-schnablelab/tree/main/src/schnablelab/GC/data">data directory</a>.</p>
<h3 id="qc-on-missing-data">QC on Missing Data</h3>
<p>This step removes SNPs and samples with a higher-than-specified missing data ratio (<code>--cutoff_snp</code> and <code>--cutoff_sample</code>).</p>
<p><div class="language-text highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>$ python -m schnablelab.GC.data_qc qc_missing RIL_X0.4miss0.35.map qc_missing.map --cutoff_snp 0.4 --cutoff_sample 0.4
</code></pre></div>
Here, <code>RIL_X0.4miss0.35.map</code> is an <a href="https://github.com/chenyong89/unl-schnablelab/blob/main/src/schnablelab/GC/data/RIL_X0.4miss0.35.map">example input genotype table</a>, and <code>qc_missing.map</code> is the resulting table after removing high missing ratio SNPs and samples.</p>
<h3 id="qc-on-snps-with-distorted-segregation-ratio">QC on SNPs with Distorted Segregation Ratio</h3>
<p>This step removes SNPs with highly distorted segregation ratios. </p>
<p><div class="language-text highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>$ python -m schnablelab.GC.data_qc qc_sd qc_missing.map qc_segregation.map --population RIL
</code></pre></div>
In this command, <code>qc_missing.map</code> is the table from the previous QC step, and <code>qc_segregation.map</code> is the resulting table.</p>
<h3 id="qc-on-homozygous-genotype-calls-in-heterozygous-regions">QC on Homozygous Genotype Calls in Heterozygous Regions</h3>
<p>This step identifies small regions (defined by read length in base pairs) where continuous homozygous genotype calls occur within a heterozygous region.</p>
<p><div class="language-text highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>$ python -m schnablelab.GC.data_qc qc_hetero qc_segregation.map qc_hetero.map --read_len 150
</code></pre></div>
Here, <code>qc_segregation.map</code> is the table from the previous QC step, and <code>qc_hetero.map</code> is the resulting table. The process also generates a log file <code>GC.qc_hetero.info</code> detailing aggregated regions.</p>
<h3 id="qc-on-duplicated-snp-markers">QC on Duplicated SNP Markers</h3>
<p>This step aggregates consecutive markers with identical genotype calls across all samples/lines.</p>
<p><div class="language-text highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>$ python -m schnablelab.GC.data_qc qc_dup qc_hetero.map qc_dup.map
</code></pre></div>
In this command, <code>qc_hetero.map</code> is the table from the previous QC step, and <code>qc_dup.map</code> is the resulting table. Use <code>--diff_num</code> to specify the number of tolerated different genotype calls between consecutive SNP markers (default is 0). Aggregation information is logged in <code>GC.qc_dup.log</code>.</p>
<h2 id="step-3-conduct-the-correction">Step 3. Conduct the Correction</h2>
<h3 id="prepare-the-configuration-file">Prepare the Configuration File</h3>
<p>The configuration file defines parameters for the correction step. An example config file can be found <a href="https://github.com/chenyong89/unl-schnablelab/blob/main/src/schnablelab/GC/data/config.txt">here</a>. You can modify it based on your own input data.
<img alt="config" src="../figures/config.png" /></p>
<h3 id="run-genotype-corrector">Run Genotype-Corrector</h3>
<p>Once you have the genotype table after QC and the configuration file, run Genotype-Corrector using:</p>
<div class="language-text highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>$ python -m schnablelab.GC.corrector correct config.txt qc_dup.map --itertimes 5
</code></pre></div>
<p>Upon completion, the corrected genotype table <code>qc_dup.corrected.map</code> will be generated where missing values are imputed and genotype calls are corrected. Adding <code>--debug</code> parameter to above command line will generate an output table <code>qc_dup.corrected.map.debug</code> containing both corrected genotype calls followed by the original ones highlighted in parentheses as shown below. </p>
<p><img alt="debug_table" src="../figures/debug.png" /></p>
<p>You can manually debug results if you found obvious wrong corrections by GC in this file. After debugging, you can run <code>python -m schnablelab.GC.corrector cleanup</code> command to remove redudant info in the table. For conversion to other formats (e.g., R/QTL or MSTmap), use <code>python -m schnable.GC.corrector format</code> for downstream analyses.</p>
<h2 id="tips">Tips</h2>
<ul>
<li>GC performs best with high-density marker datasets from high-throughput sequencing, where each contig should have at least hundreds of markers.</li>
<li>Some missing values and genotype call corrections may not be performed by GC without strong supporting evidence.</li>
<li>Grid search of input parameters (e.g., sliding window size) is recommended to achieve optimized results.</li>
</ul>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var tab,labels=set.querySelector(".tabbed-labels");for(tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.tabs.link", "navigation.tabs"], "search": "../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML"></script>
      
    
  </body>
</html>